### PR TITLE
Private endpoint list

### PR DESCRIPTION
**Pull request by su-chang**

### CODE REVIEW

**Comment by su-chang** [comment]
Maybe need add method `conversationRead` to PuppetMock, or lock `wechaty-puppet` version.

**Comment by huan** [comment]
Please file an issue and descript what you plan to achieve with this PR in detail.

Thank you very much.

**Comment by su-chang** [comment]
Related issue: https://github.com/wechaty/wechaty-puppet-service/issues/141

**Comment by su-chang** [comment]
@huan ping

**Comment by huan** [comment]
Thanks for ping me, and sorry for the late reply due to the heavy load for my past days.

I believe this improvement is very valuable and important, will review it this week.

Appreciate for the effort from you for pushing this forward.

File: src/config.ts [hunk 1]
```diff
@@ -51,6 +51,18 @@ export const WECHATY_PUPPET_SERVICE_ENDPOINT = () => {
   return undefined
 }
 
+export const GET_CHATIE_ENDPOINT_LIST = () => {
+  const privateEndpointList = process.env.PRIVATE_ENDPOINT_LIST // e.g. https://aaa.aaa,http://bbb.bbb
```
Comment (windmemory) [inline src/config.ts:55]: Actually this is a list of service discovery endpoints. `PRIVATE_ENDPOINT_LIST` seems a little bit vague. So how about changing this env var to `SERVICE_DISCOVERY_ENDPOINT_LIST`? Or `SERVICE_DISCOVERY_ENDPOINTS`?

**Review by windmemory — APPROVED** [review]
Looks good to me, approved!

**Comment by su-chang** [comment]
> it will be great if you can also file an issue for documenting this PR with some clear explanation.

Related issue: https://github.com/wechaty/wechaty-puppet-service/issues/142

Please comments for improve it if you are available, thank you.

File: package.json [hunk 2]
```diff
@@ -54,7 +54,7 @@
     "shx": "^0.3.3",
     "tstest": "^0.4.10",
     "wechaty": "^0.56",
-    "wechaty-puppet": "^0.37.1",
+    "wechaty-puppet": "0.37.1",
```
Comment (huan) [inline package.json:57]: Why change this version to be locked?

File: src/client/puppet-service.ts [hunk 3]
```diff
@@ -823,6 +820,15 @@ export class PuppetService extends Puppet {
     )(request)
   }
 
+  /**
+   *
+   * Conversation
+   *
+   */
+  public async conversationRead (conversationId: string) {
```
Comment (huan) [inline src/client/puppet-service.ts:828]: Please remove unrelated changes.

If you have to add this method, I don't think it's right to add an empty function body.

You should throw an error to let users know this method has not been implemented.

File: src/config.ts [hunk 4]
```diff
@@ -51,6 +51,18 @@ export const WECHATY_PUPPET_SERVICE_ENDPOINT = () => {
...
 }
 
+export const GET_CHATIE_ENDPOINT_LIST = () => {
+  const serviceDiscoveryEndpointList = process.env.SERVICE_DISCOVERY_ENDPOINT_LIST // e.g. https://aaa.aaa,http://bbb.bbb
+  if (serviceDiscoveryEndpointList) {
+    return serviceDiscoveryEndpointList.split(',')
+  }
+  return [
+    'https://api.chatie.io',
+    'https://chatieio.herokuapp.com',
```
Comment (huan) [inline src/config.ts:61]: Please remove this one, because it's deprecated. (not worked any more)

File: src/config.ts [hunk 5]
```diff
@@ -51,6 +51,18 @@ export const WECHATY_PUPPET_SERVICE_ENDPOINT = () => {
...
+    'http://68.79.16.140',  // from @windmemory,
```
Comment (huan) [inline src/config.ts:62]: Could we use a DNS name for this instead of the IP?

For example, chatie-api-proxy.api.juzi.bot with `https` enabled?

File: src/config.ts [hunk 6]
```diff
@@ -51,6 +51,18 @@ export const WECHATY_PUPPET_SERVICE_ENDPOINT = () => {
   return undefined
 }
 
+export const GET_CHATIE_ENDPOINT_LIST = () => {
```
Comment (huan) [inline src/config.ts:54]: I think we should not use a list here anymore.

When we add this **list** to our code, the reasons are: some server has issues visiting the Cloudflare/Heroku IP (slow or even timeout).

After we can override this endpoint, then the user who has the above issue will be able to use their private endpoint to solve the problem.

So I'd like to suggest that we only keep one endpoint here.

Please let me know what you think @windmemory @su-chang

**Review by huan — CHANGES_REQUESTED** [review]

File: src/config.ts (hunk 6 above)
Comment (su-chang) [reply src/config.ts:54]: The default endpoint still keep `[ 'https://api.chatie.io', 'http://68.79.16.140' ]`, and keep only one endpoint is for private endpoint?

File: package.json (hunk 2 above)
Comment (su-chang) [reply package.json]: **log**
```log
> wechaty-puppet-service@0.17.2 lint:ts /home/runner/work/wechaty-puppet-service/wechaty-puppet-service
> tsc --noEmit

Error: tests/performance.spec.ts(37,7): error TS2515: Non-abstract class 'PuppetTest' does not implement inherited abstract member 'conversationRead' from class 'PuppetMock'.
```
I do not know how to fix this problem...

Comment (windmemory) [reply package.json]: You have to add `conversationRead` method to the class called `PuppetTest`. This class should be in the test folder.

Comment (su-chang) [reply package.json]: After `conversationRead` been added, it could pass the test, but it is strange that only two methods in class `PuppetTest`.

File: src/config.ts (hunk 6 above)
Comment (huan) [reply src/config.ts]: I'm considering only keep one default endpoint, but it's not decided yet.

Comment (windmemory) [reply src/config.ts]: Vote for one default endpoint.

Multiple endpoints was introduced to resolve network latency issue, and seems like we don't have that problem anymore, so we can change it back to single endpoint.

**Review by windmemory — APPROVED** [review]
Looks good to me, approved!

**Comment by windmemory** [comment]
ping @huan 

Any other comments?

**Comment by windmemory** [comment]
@huan kindly ping

**Review by huan — APPROVED** [review]
LGTM

**Comment by huan** [comment]
Thanks for this improvement.

I have renamed the variable name to `WECHATY_SERVICE_DISCOVERY_ENDPOINT`, please try it (0.21.3 or above) and let me know if there's any problem.
